Add support to zero-copy Kaldi CuVector/CuMatrix to PyTorch's GPU tensor.

csukuangfj opened the pull request:
Actually, it can zero-copy Kaldi's CuVector/CuMatrix to
any DL frameworks that support the DLPack interface.

CuVector/Vector/CuMatrix/Matrix support ONLY `to_dlpack`.

Cu**Sub**Vector/**Sub**Vector/Cu**Sub**Matrix/**Sub**Matrix support
both `from_dlpack` and `to_dlpack`

csukuangfj committed:
Add support to zero-copy Kaldi CuVector/CuMatrix to PyTorch's GPU
tensor.

Actually, it can zero-copy Kaldi's CuVector/CuMatrix to
any DL frameworks that support the DLPack interface.

csukuangfj commented:
Next, we will update our internal LF-MMI training pipleline with kaldi pybind instead of PyKaldi.

Everything will in kaldi pybind.

csukuangfj commented:
@danpovey @naxingyu @cweng6
please have a review.

danpovey commented on src/pybind/cudamatrix/cu_matrix_pybind.cc:
```diff
@@ -37,10 +38,71 @@ void pybind_cu_matrix(py::module& m) {
+             })
+#if HAVE_CUDA == 1
+        .def("to_dlpack",
+             [](PyClass* m) {
```
Bear in mind that if HAVE_CUDA is not 1, it will actually contain CPU memory.  You might want to eventually handle that case.

csukuangfj replied:
My intention is that if a user in Python wants to use CuMatrix/CuVector, then he/she means the GPU one. If he wants only CPU, then he/she should use Matrix/Vector.

csukuangfj replied:
Now it will call KALDI_ERR for `to_dlpack` if kaldi is not compiled with CUDA but the user
calls `to_dlpack` for CuVectorBase/CuMatrixBase.

danpovey commented on src/pybind/matrix/matrix_pybind.cc:
```diff
@@ -49,12 +50,56 @@ void pybind_matrix(py::module& m) {
+        // we use the name `to_dlpack` because PyTorch uses the same name
+
+        // the created `managed_tensor` will be freed in
+        // `DLManagedTensorDeleter`, so no memory leak here
```
change to
```
`DLManagedTensorDeleter`, which does not free `data`, so no memory leak here`
```

csukuangfj replied:
Will fix it in the next patch.

csukuangfj replied:
Done.

danpovey commented on src/pybind/matrix/vector_pybind.cc:
```diff
@@ -44,20 +45,66 @@ void pybind_vector(py::module& m) {
+             // using  the  method `numpy()` to convert a matrix/tensor to a
+             // numpy array.
+           })
+      .def("to_dlpack", [](VectorBase<float>* v) {
```
I think now is the best time to add documentation for this, while it's still fresh in your mind.

csukuangfj replied:
do we need to create a directory`pybind/doc` and puts a `kaldi_pybind.dox` in it ?

naxingyu commented on src/pybind/dlpack/dlpack_pybind_test.py:
```diff
@@ -108,6 +109,156 @@ def test_pytorch_gpu_tensor_to_cu_subvector_and_cu_submatrix(self):
+        gc.collect()
+
+        # one more time
+
```
extra newline?

csukuangfj replied:
Will fix in the next patch.

csukuangfj replied:
Done

danpovey commented on src/pybind/matrix/vector_pybind.cc:
```diff
@@ -44,20 +45,66 @@ void pybind_vector(py::module& m) {
       });
 
+  // TODO(fangjun): add `from_dlpack` once kaldi::Vector is changed to borrow a
+  // ptr from the outside
```
@csukuangfj the issue with adding from_dlpack to Vector is it won't work for CuVector due to different memory allocators.  Would be better, I think, to add it to SubVector, and ideally somehow have the memory be managed by the Python object?  [when we wrap these things, I assume there must be some kind of Python object that contains the C++ object?]

csukuangfj replied:
I agree.

csukuangfj replied:
now SubVector/SubMatrix/CuSubVector/CuSubMatrix support both
`from_dlpack` and `to_dlpack`.

Vector/Matrix/CuVector/CuMatrix support ONLY `to_dlpack`

cweng6 commented on src/pybind/matrix/matrix_pybind.cc:
```diff
@@ -49,12 +50,56 @@ void pybind_matrix(py::module& m) {
-            {sizeof(float) * m->Stride(), sizeof(float)},  // stride in bytes
-            m->Data(),                                     // ptr
-            py::none());  // pass a base object to avoid copy!
+      .def("numpy",
```
For CPU kaldi matrix / pytorch tensor zero-copy, do we really need to convert to dlpack since one line of binding code,

.def_buffer([](Matrix<float> &mat) -> py::buffer_info {
      return py::buffer_info(
        mat.Data(), /*raw pointer*/
        sizeof(float),/*size of one scalar*/
        py::format_descriptor<float>::format(),
        2, /*number of dimensions*/
        {mat.NumRows(), mat.NumCols()} , /*buffer dimension*/
        {sizeof(float) * mat.Stride(), sizeof(float)} /*strides in bytes*/
      );

will accomplish zero-copy between kaldi matrix  and numpy array. And I believe all DL framework support numpy array now. therefore using numpy as a bridge instead of dlpack is more straghtforward.

csukuangfj replied:
Yes, you're right.

This method `numpy` does not use DLPack interface. As far as I know, numpy does not support DLPack.

The method you mentioned allows us to convert a `kaldi::Vector<float>` to `numpy`
by using

```python
numpy_array = np.array(kaldi_vector, copy=False)
```

By adding this method, we can use

```python
numpy_array = kaldi_vector.numpy()
```
which I think is much easier and much more convenient.

cweng6 replied:
it might be more straightforward to call numpy(), but not sure it's easier/more convenient. 

I thought DLPack, is just an interface to allow you to accomplish zero-copy. Why do we need to support DLPack if we could do zero-copy w/o it?

csukuangfj replied:
If I understand you correctly, you think that `DLPack` is used to convert a kaldi vector to or from numpy ndarray.

The current usage of DLPack is to zero-copy data from Kaldi to some framework, 
or zero copy-data from some framework to Kaldi, both for CPU and GPU data, as
long as the framework supports the DLPack interface.

DLPack is not used for numpy.  From C++ to Python, DLPack wraps a data ptr
and some other meta information in it and is sent as a PyCapsule to Python.
In Python, if a framework supports DLPack interface, it will consume this PyCapsule
and extracts the data ptr and the meta-information to construct a tensor specific
to this framework. 

I think DLPack has nothing to do with numpy. It is mainly used for zero-copying
data between different frameworks. Numpy does not support DLPack interface,
since we cannot construct a numpy ndarray directly from a PyCapsule.

csukuangfj replied:
DLPack can also support zero-copying **GPU** data.

csukuangfj replied:
I prefer
```python
numpy_array = kaldi_vector.numpy()
```

to 

```python
numpy_array = np.array(kaldi_vector, copy=False)
```

PyKaldi and PyTorch also support this kind of syntax `numpy_array = tensor.numpy()`

cweng6 replied:
I didn't mean using DLPack is to zero-copy between Kaldi vector and numpy and was aware of that DLPack is irrelevant to numpy. I was saying since all DL framework supports numpy, and with numpy using buffer protocol, we could easily zero-copying CPU data between different frameworks why bother using DLpack at all for CPU?

csukuangfj replied:
I see and I agree with you.

`kaldi <--> numpy <---> some DL framework` is enough for **CPU** data
and we do NOT need `kaldi <---> DLPack <---> some DL framework`

But with DLPack, we can handle CPU/GPU data homogeneously:

For example, to zero copy data from kaldi to some DL framework, the following function is enough

```python
def from_kaldi_to_some_dl_framework(v):
  tensor = from_dlpack(v.to_dlpack())
```

Without DLPack, we have to do
```python
def from_kaldi_to_some_dl_framework(v):
    if v is on GPU:
      tensor = from_dlpack(v.to_dlpack())
    else:
      tensor = some_framework.from_numpy(v.numpy())
```

cweng6 replied:
Make much sense, this way zero-copy can be device-agostic

csukuangfj replied:
😄

csukuangfj replied:
Mark as resolved.

cweng6 commented on src/pybind/matrix/vector_pybind.cc:
```diff
@@ -44,20 +45,66 @@ void pybind_vector(py::module& m) {
-        // `numpy` instead of `Numpy`, `ToNumpy` or `SomeOtherName` is used here
-        // because we want to follow the style in PyKaldi and PyTorch using the
-        // method `numpy()` to convert a matrix/tensor to a numpy array.
+      .def("numpy",
```
similar thing for CPU vectors, see my previous comments, using numpy array as an interface with one line of binding suffices

csukuangfj replied:
See above. Mark as resolved.

cweng6 reviewed (commented):
overall LGTM. As I mentioned, for CPU kaldi matrix/vectors, I prefer using numpy (buffer protocol binding) instead of DLpack.

danpovey commented:
Let's keep the code for now, it will help familiarize us with DLPack.  We
can delete it later if we still feel it's not needed.


On Mon, Dec 23, 2019 at 2:47 PM cweng6 <notifications@github.com> wrote:

> *@cweng6* commented on this pull request.
> ------------------------------
>
> In src/pybind/matrix/matrix_pybind.cc
> <https://github.com/danpovey/kaldi/pull/93#discussion_r360782759>:
>
> > @@ -49,12 +50,56 @@ void pybind_matrix(py::module& m) {
>             [](MatrixBase<float>& m, std::pair<ssize_t, ssize_t> i, float v) {
>               m(i.first, i.second) = v;
>             })
> -      .def("numpy", [](MatrixBase<float>* m) {
> -        return py::array_t<float>(
> -            {m->NumRows(), m->NumCols()},                  // shape
> -            {sizeof(float) * m->Stride(), sizeof(float)},  // stride in bytes
> -            m->Data(),                                     // ptr
> -            py::none());  // pass a base object to avoid copy!
> +      .def("numpy",
>
> I didn't mean using DLPack is to zero-copy between Kaldi vector and numpy
> and was aware of that DLPack is irrelevant to numpy. I was saying since all
> DL framework supports numpy, and with numpy using buffer protocol, we could
> easily zero-copying CPU data between different frameworks why bother using
> DLpack at all for CPU?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/danpovey/kaldi/pull/93?email_source=notifications&email_token=AAZFLO4ONNJT6I74SSWTGYTQ2BNGJA5CNFSM4J6PS6BKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCQBNP5Y#discussion_r360782759>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLOZPYKH72KQSYUAAEZDQ2BNGJANCNFSM4J6PS6BA>
> .
>

danpovey commented:
I was just talking about the code-level documentation, whatever it is that
you have to put in there so the documentation shows up in python
appropriately.
We can think later about doxygen-level documentation.  (or equivalents.)


On Mon, Dec 23, 2019 at 4:18 PM Fangjun Kuang <notifications@github.com>
wrote:

> *@csukuangfj* commented on this pull request.
> ------------------------------
>
> In src/pybind/matrix/vector_pybind.cc
> <https://github.com/danpovey/kaldi/pull/93#discussion_r360802071>:
>
> > +           [](VectorBase<float>* v) {
> +             return py::array_t<float>(
> +                 {v->Dim()},       // shape
> +                 {sizeof(float)},  // stride in bytes
> +                 v->Data(),        // ptr
> +                 py::none());      // pass a base object to avoid copy!
> +             // (fangjun): the base object can be anything containing a non-null
> +             // ptr. I cannot think of a better way than to pass a `None`
> +             // object.
> +             //
> +             // `numpy` instead of `Numpy`, `ToNumpy` or `SomeOtherName` is used
> +             // here because we want to follow the style in PyKaldi and PyTorch
> +             // using  the  method `numpy()` to convert a matrix/tensor to a
> +             // numpy array.
> +           })
> +      .def("to_dlpack", [](VectorBase<float>* v) {
>
> do we need to create a directorypybind/doc and puts a kaldi_pybind.dox in
> it ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/danpovey/kaldi/pull/93?email_source=notifications&email_token=AAZFLO7WX5XWQ4EARRUUAALQ2BX4NA5CNFSM4J6PS6BKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCQBTGLQ#discussion_r360802071>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLO644V6WKDO6LJKFZ6TQ2BX4NANCNFSM4J6PS6BA>
> .
>

csukuangfj commented:
`CuPy` supports conversion not only to Numpy, but also to `DLPack` and to some other data structure.

See https://docs-cupy.chainer.org/en/stable/reference/interoperability.html#dlpack

If we use DLPack to zero-copy GPU data, why not to support zero-copying CPU data?

---
TVM also supports both numpy and DLPack, see

- https://docs.tvm.ai/api/python/ndarray.html#tvm.ndarray.NDArray.asnumpy
- https://docs.tvm.ai/api/python/ndarray.html#tvm.ndarray.NDArray.to_dlpack

Why other frameworks support both when only `numpy` is enough?

csukuangfj commented:
For the documentation, I will dig into Open3D to borrow some ideas.

danpovey commented:
OK... I prefer documentation that is inline with the code, either comments
or doc-string types of things.
Documentation that's separated from the code will have a problem with
discoverability.
Just comments might be enough.  But someone mentioned that there is a way
to add strings to
pybind11 so that it shows up from the python command line as documentation
for the relevant
functions.

On Mon, Dec 23, 2019 at 4:25 PM Fangjun Kuang <notifications@github.com>
wrote:

> For the documentation, I will dig into Open3D to borrow some ideas.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/danpovey/kaldi/pull/93?email_source=notifications&email_token=AAZFLOYAMUCIWQVSKQY5UIDQ2BYWDA5CNFSM4J6PS6BKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHQRTEI#issuecomment-568400273>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLO2YGJ3BJDAQPMB7ALDQ2BYWDANCNFSM4J6PS6BA>
> .
>

csukuangfj commented:
I see. I guess I suggested this. I will look into Open3D.

csukuangfj committed:
add from_dlpack to SubVector/SubMatrix/CuSubVector/CuSubMatrix.

csukuangfj commented:
@danpovey 
I think I should finish LF-MMI training first and will do documentation after that.

danpovey commented:
Fantastic!  I am amazed with your speed.  I'll merge this.  You can make the next PR on kaldi's own account.  BTW, Meixu added the slack app to kaldi's github account, we will try to figure out how it works (and github's "projects" feature, which seems to be connected to it)

csukuangfj commented:
ok, we will go to the official GitHub of kaldi for the next pullrequest.

csukuangfj commented:
@danpovey 
I guess there will be some groups in Slack, could you please add us to it?

danpovey commented:
Meixu is managing that.  I got a link from him but it said it expired, I
think he is working on it.


On Mon, Dec 23, 2019 at 6:59 PM Fangjun Kuang <notifications@github.com>
wrote:

> @danpovey <https://github.com/danpovey>
> I guess there will be some groups in Slack, could you please add us to it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/danpovey/kaldi/pull/93?email_source=notifications&email_token=AAZFLO5XS6DLSLTDRX4AGWDQ2CK2PA5CNFSM4J6PS6BKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHQ35OQ#issuecomment-568442554>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLO7CPJ6X4PUC7IL5A3LQ2CK2PANCNFSM4J6PS6BA>
> .
>

megazone87 commented:
I am up to it right now. I am adding some cool plugins into it. Wait a moment please.
2019年12月23日 +0800 18:59 Fangjun Kuang <notifications@github.com>，写道：

@danpovey<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdanpovey&data=02%7C01%7C%7C0a47f63e2cb84673fac508d787973c59%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637126955930700706&sdata=B6%2Bm46Zd3d%2BAjxGnxZ5%2FUbBkgTrOMJYl9yJdYcrftdQ%3D&reserved=0>
I guess there will be some groups in Slack, could you please add us to it?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdanpovey%2Fkaldi%2Fpull%2F93%3Femail_source%3Dnotifications%26email_token%3DAAFMILWWE5PHB4KLR2TRHODQ2CK2PA5CNFSM4J6PS6BKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHQ35OQ%23issuecomment-568442554&data=02%7C01%7C%7C0a47f63e2cb84673fac508d787973c59%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637126955930700706&sdata=08fwyBBpHXhYxVcO3MLuMo5tBPwBMDnOtfdoXhDGJYo%3D&reserved=0>, or unsubscribe<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAAFMILRCOLOSJMVBVZXF6GTQ2CK2PANCNFSM4J6PS6BA&data=02%7C01%7C%7C0a47f63e2cb84673fac508d787973c59%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637126955930710711&sdata=uk0DOhuX4eeZn5tb%2F%2F7pJpuwjXmGqsOTaxfjNj4gTFc%3D&reserved=0>.

csukuangfj commented:
all right. we'll wait.

csukuangfj commented:
It happens that there exists **potential** memory leak for DLPack. I'll fix it and write the details as soon as possible.

naxingyu committed:
Add online feature (#3783)

* add some of the online features

* update

* resolve most comments

* add Matrix<double> bind and OnlineCmvnState

csukuangfj committed:
fix potential memory leak in DLPack.

In the following, we use two notations:

- producer: the one who produces DLManagedTensor
- consumer: the one who consumes DLManagedTensor

When to zero-copy data between C++ and Python, the common practice
used by PyTorch, CuPy and TVM is to use a PyCapsule.

The producer sets the name of the PyCapsule to `dltensor`
and the consumer rewrites the name to `used_dltensor`.

When the producer finds that the PyCapsule has no consumer
by comparing its name, the producer is responsible to free
the memory of `DLManagedTensor`;

When the consumer uses a PyCapsule, it is the responsibility
of the consumer to free the memory of `DLManagedTensor`.

Note that a PyCapsule can only be used by ONE consumer.

csukuangfj commented:
Why was there a potential memory leak in the previous commit?

When we create a `DLManagedTensor`, we assume implicitly that it will be used
by PyTorch and PyTorch will free it. But the problem is that if the user creates
a `DLManagedTensor` and does not pass it to PyTorch, then the resource is lost.

Let's define two terms:

- producer: the one who creates the DLManaged tensor and wraps it to a PyCapsule
- consumer: the one who consumes the PyCapsule and extracts the data from the DLManagedTensor

When the producer produces the PyCapsule, I guess it is a convention to give it a name `dltensor`;
when the consumer consumes the PyCapsule, the consumer rewrites the name to `used_dltensor`.

By comparing against the name, the producer will know whether a PyCapsule is consumed;
if not, it is the responsibility of the producer to free any resources inside DLManagedTensor.

If a consumer consumes a PyCasule, the consumer MUST free the resource.

Note that a PyCapsule can ONLY be used by one consumer.

---

There are many issues relating to the memory leak problems of DLPack, just google it.

One is `memory leak for PyTorch -> dlpack -> CuPy` (https://github.com/pytorch/pytorch/issues/9261)

I reproduce the code for kaldi pybind here

```python
import os
import sys
sys.path.insert(0, os.path.join(os.path.dirname(__file__), os.pardir))

import gc

try:
    import torch
except ImportError:
    print('This test needs PyTorch.')
    print('Please install PyTorch first.')
    print('PyTorch 1.3.0dev20191006 has been tested and is guaranteed to work.')
    sys.exit(0)

from torch.utils.dlpack import to_dlpack

import kaldi

device_id = 0
kaldi.SelectGpuDevice(device_id=device_id)

device = torch.device('cuda', device_id)
for i in range(10000):
    b = torch.randn(1000, 1000, device=device)
    a = kaldi.CuSubMatrixFromDLPack(torch.utils.dlpack.to_dlpack(b))
    gc.collect()
```

`watch -n 0.5 nvidia-smi` shows that the used memory is a constant, i.e., 661 MB out of 12196 MB.

csukuangfj commented:
@danpovey 
should I close this pullrequest and reopen it at kaldi-asr/kaldi

danpovey commented:
Thanks a lot!  Yes please re-open there.  And thanks for the explanation.
What I am curious about, though, is how frameworks ensure that this memory is only used in one place and how the deletion is managed.  Because normally there will be 
  (a) some kind of memory manager, since CUDA's one is slow
  (b) likely a shared_ptr type of arrangement so multiple objects can safely refer to the allocated region,
  (c) arrays may actually be referring to sub-parts of an allocated region

So I'm wondering whether dlpack requires that you just allocate a memory area using plain old cuMalloc()?

csukuangfj commented:
Everything about DLPack is contained in this single header file: https://github.com/dmlc/dlpack/blob/master/include/dlpack/dlpack.h

DLPack does not specify how you allocate the memory. It just wraps a data pointer
and some meta information about it. It can optionaly specify a deleter (https://github.com/dmlc/dlpack/blob/9d0d02b9a6fc352da1dc223512a23a0d8c87d056/include/dlpack/dlpack.h#L168), which is responsible to free any resources.

The information passed between frameworks via DLPack is just the data pointer
and the meta information, i.e., strides, shapes, offset.

DLPack does not specify how you should implement it.
DLPack is created by Yangqing jia (https://github.com/Yangqing), the creator of `Caffe`. I guess you know him or have heard his name. 

He also implemented DLPack in PyTorch and Caffe2 and used `dltensor` and `used_dltensor` as 
fixed names for PyCapsule. If other frameworks want to communicate with PyTorch, they
must follow the same convention.

The framework cannot ensure that the data pointer in the DLManagedTensor
is used in one place.  If a consumer wants to use a PyCasule, he MUST first
check the name of the PyCasule, if it is not `dltensor`, then it throws;
otherwise, it set the name to `used_dltensor` to prevent other consumers using it.

As for the deletion, If there is no consumer, when the PyCapsule is going to be collected
by the garbage collector, the destructor of the Pycapsule will call the deleter.
When there is a consumer,  the consumer MUST ensure that when the object created
from the PyCapsule goes out of scope, the deleter is called.

That's why I created a new class `DLPackSubVector` which saves the pointer
to the `DLManagedTensor`. When  `DLPackSubVector` goes out of scope, we
invoke the deleter in its destructor.

csukuangfj commented:
close this now.
reopend at https://github.com/kaldi-asr/kaldi/pull/3788

csukuangfj committed:
move `to_dlpack` from MatrixBase to Matrix.
